[css-shapes] Implement xywh() shape

nmoucht opened the pull request:
#### 6dd9b4ce08d7fc1933952dfd57de080a21882127
<pre>
[css-shapes] Implement xywh() shape
<a href="https://bugs.webkit.org/show_bug.cgi?id=259989">https://bugs.webkit.org/show_bug.cgi?id=259989</a>
rdar://113643020

Reviewed by Tim Nguyen.

Implement parsing of xywh() shape and create xywh
subclass for BasicShape and CSSValue.

Translating to inset() for the computed style &amp; animations is unimplemented.

Spec: <a href="https://drafts.csswg.org/css-shapes-1/#funcdef-basic-shape-xywh">https://drafts.csswg.org/css-shapes-1/#funcdef-basic-shape-xywh</a>

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/animations/clip-path-interpolation-xywh-rect-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/parsing/clip-path-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/parsing/clip-path-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/motion/animation/offset-path-interpolation-006-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/motion/parsing/offset-path-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/motion/parsing/offset-path-parsing-valid-expected.txt:
* Source/WebCore/css/BasicShapeFunctions.cpp:
(WebCore::valueForBasicShape):
(WebCore::basicShapeForValue):
* Source/WebCore/css/CSSBasicShapes.cpp:
(WebCore::CSSXywhValue::CSSXywhValue):
(WebCore::CSSXywhValue::create):
(WebCore::CSSXywhValue::equals const):
(WebCore::updateCornerRadiusWidthAndHeight):
(WebCore::buildRadii):
(WebCore::buildXywhString):
(WebCore::CSSXywhValue::customCSSText const):
(WebCore::buildInsetString):
(WebCore::buildInsetRadii): Deleted.
* Source/WebCore/css/CSSBasicShapes.h:
(WebCore::CSSXywhValue::insetX const):
(WebCore::CSSXywhValue::insetY const):
(WebCore::CSSXywhValue::width const):
(WebCore::CSSXywhValue::height const):
(WebCore::CSSXywhValue::topLeftRadius const):
(WebCore::CSSXywhValue::topRightRadius const):
(WebCore::CSSXywhValue::bottomRightRadius const):
(WebCore::CSSXywhValue::bottomLeftRadius const):
* Source/WebCore/css/CSSValue.cpp:
(WebCore::CSSValue::visitDerived):
* Source/WebCore/css/CSSValue.h:
(WebCore::CSSValue::isXywhShape const):
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeShapeBorderRadius):
(WebCore::CSSPropertyParserHelpers::consumeBasicShapeXywh):
(WebCore::CSSPropertyParserHelpers::consumeBasicShapeInset):
(WebCore::CSSPropertyParserHelpers::consumeBasicShape):
* Source/WebCore/rendering/style/BasicShapes.cpp:
(WebCore::BasicShapeXywh::create):
(WebCore::BasicShapeXywh::BasicShapeXywh):
(WebCore::BasicShapeXywh::clone const):
(WebCore::BasicShapeXywh::operator== const):
(WebCore::BasicShapeXywh::path):
(WebCore::BasicShapeXywh::canBlend const):
(WebCore::BasicShapeXywh::blend const):
(WebCore::BasicShapeXywh::dump const):
* Source/WebCore/rendering/style/BasicShapes.h:

Canonical link: <a href="https://commits.webkit.org/267789@main">https://commits.webkit.org/267789@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f2a9ab89901d887b3d9198a8c33f123e780addad

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/17707 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/18037 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/18573 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/19532 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/16563 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/17902 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/21328 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/18186 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/18629 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/17920 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/18216 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/15385 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/20393 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/15450 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/16134 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/22706 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/16460 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/16304 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/20565 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/16870 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/14287 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/15986 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4218 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/20349 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/16715 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/CSSBasicShapes.cpp:
```diff
@@ -191,6 +191,118 @@ bool CSSEllipseValue::equals(const CSSEllipseValue& other) const
+    const String& bottomLeftRadiusWidth, const String& bottomLeftRadiusHeight)
+{
+    StringBuilder result;
+    result.append("xywh(", offsetX, offsetY, width, height);
```
```suggestion
    result.append("xywh("_s, offsetX, ' ', offsetY, ' ', width, ' ', height);
```

nt1m commented on Source/WebCore/css/CSSBasicShapes.cpp:
```diff
@@ -191,6 +191,118 @@ bool CSSEllipseValue::equals(const CSSEllipseValue& other) const
+}
+
+
+
```
nit: excess new lines.

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -6954,6 +6954,43 @@ bool consumeRadii(std::array<RefPtr<CSSValue>, 4>& horizontalRadii, std::array<R
+static std::optional<std::array<RefPtr<CSSValuePair>, 4>> consumeBorderRadius(CSSParserTokenRange& args, const CSSParserContext& context)
+{
+    std::array<RefPtr<CSSValuePair>, 4> radii;
+    if (consumeIdent<CSSValueRound>(args)) {
```
```suggestion
    if (consumeIdentRaw<CSSValueRound>(args)) {
```
Since you don't need the CSSValue, consumeIdentRaw will be faster.

Also writing this as an early return would be nice:

```
if (!consumeIdentRaw<CSSValueRound>(args))
  return std::nullopt;
```

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -6954,6 +6954,43 @@ bool consumeRadii(std::array<RefPtr<CSSValue>, 4>& horizontalRadii, std::array<R
     return true;
 }
 
+static std::optional<std::array<RefPtr<CSSValuePair>, 4>> consumeBorderRadius(CSSParserTokenRange& args, const CSSParserContext& context)
```
Can we name this `consumeBorderRadiusForShape` ? or share with the border-radius parsing?

nt1m commented on Source/WebCore/rendering/style/BasicShapes.cpp:
```diff
@@ -324,6 +324,103 @@ void BasicShapeEllipse::dump(TextStream& ts) const
+
+bool BasicShapeXywh::canBlend(const BasicShape& other) const
+{
+    return (type() == other.type());
```
nit: extra brackets

nt1m reviewed (commented):
Do any of the clip-path WPT Mozilla added pass? https://wpt.fyi/results/css/css-masking?label=master&label=experimental&aligned&q=xywh

darinadler commented on Source/WebCore/css/CSSBasicShapes.cpp:
```diff
@@ -191,6 +191,116 @@ bool CSSEllipseValue::equals(const CSSEllipseValue& other) const
+            for (auto& radius : horizontalRadii)
+                result.append(' ', radius);
+
+            if (verticalRadii.size() != horizontalRadii.size()
```
Why can't this just be `if (verticalRadii != horizontalRadii)`? `VectorComparer` is supposed to be an implementation detail of `Vector` and I am surprised you need to use it directly.

darinadler commented on Source/WebCore/css/CSSBasicShapes.cpp:
```diff
@@ -191,6 +191,116 @@ bool CSSEllipseValue::equals(const CSSEllipseValue& other) const
+    String bottomLeftRadiusWidth;
+    String bottomLeftRadiusHeight;
+
+    updateCornerRadiusWidthAndHeight(topLeftRadius(), topLeftRadiusWidth, topLeftRadiusHeight);
```
It would be much more efficient to find a way to build one string without intermediate strings.

nt1m replied:
This code shared with inset(), so cleaning up would be better as a followup.

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -6978,6 +6978,41 @@ bool consumeRadii(std::array<RefPtr<CSSValue>, 4>& horizontalRadii, std::array<R
+static RefPtr<CSSXywhValue> consumeBasicShapeXywh(CSSParserTokenRange& args, const CSSParserContext& context)
+{
+    std::array<RefPtr<CSSValue>, 2> insets;
+    for (unsigned i = 0; i < 2; ++i) {
```
This could be a range based for loop instead of hard coding the size.
```
for (auto& inset : insets) {
    inset = consumeLengthOrPercent (args, context.mode, ValueRange::A11);
    if (!inset)
        return nullptr;
}
```

nmoucht committed:
[css-shapes] Implement xywh() shape
https://bugs.webkit.org/show_bug.cgi?id=259989
rdar://113643020

Reviewed by Tim Nguyen.

Implement parsing of xywh() shape and create xywh
subclass for BasicShape and CSSValue.

Translating to inset() for the computed style & animations is unimplemented.

Spec: https://drafts.csswg.org/css-shapes-1/#funcdef-basic-shape-xywh

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/animations/clip-path-interpolation-xywh-rect-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/parsing/clip-path-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/parsing/clip-path-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/motion/animation/offset-path-interpolation-006-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/motion/parsing/offset-path-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/motion/parsing/offset-path-parsing-valid-expected.txt:
* Source/WebCore/css/BasicShapeFunctions.cpp:
(WebCore::valueForBasicShape):
(WebCore::basicShapeForValue):
* Source/WebCore/css/CSSBasicShapes.cpp:
(WebCore::CSSXywhValue::CSSXywhValue):
(WebCore::CSSXywhValue::create):
(WebCore::CSSXywhValue::equals const):
(WebCore::updateCornerRadiusWidthAndHeight):
(WebCore::buildRadii):
(WebCore::buildXywhString):
(WebCore::CSSXywhValue::customCSSText const):
(WebCore::buildInsetString):
(WebCore::buildInsetRadii): Deleted.
* Source/WebCore/css/CSSBasicShapes.h:
(WebCore::CSSXywhValue::insetX const):
(WebCore::CSSXywhValue::insetY const):
(WebCore::CSSXywhValue::width const):
(WebCore::CSSXywhValue::height const):
(WebCore::CSSXywhValue::topLeftRadius const):
(WebCore::CSSXywhValue::topRightRadius const):
(WebCore::CSSXywhValue::bottomRightRadius const):
(WebCore::CSSXywhValue::bottomLeftRadius const):
* Source/WebCore/css/CSSValue.cpp:
(WebCore::CSSValue::visitDerived):
* Source/WebCore/css/CSSValue.h:
(WebCore::CSSValue::isXywhShape const):
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeShapeBorderRadius):
(WebCore::CSSPropertyParserHelpers::consumeBasicShapeXywh):
(WebCore::CSSPropertyParserHelpers::consumeBasicShapeInset):
(WebCore::CSSPropertyParserHelpers::consumeBasicShape):
* Source/WebCore/rendering/style/BasicShapes.cpp:
(WebCore::BasicShapeXywh::create):
(WebCore::BasicShapeXywh::BasicShapeXywh):
(WebCore::BasicShapeXywh::clone const):
(WebCore::BasicShapeXywh::operator== const):
(WebCore::BasicShapeXywh::path):
(WebCore::BasicShapeXywh::canBlend const):
(WebCore::BasicShapeXywh::blend const):
(WebCore::BasicShapeXywh::dump const):
* Source/WebCore/rendering/style/BasicShapes.h:

Canonical link: https://commits.webkit.org/267789@main
